galley/pkg/mcp/source.go

```diff
@@ -83,45 +124,79 @@ func (s *source) Stop() {
 }
 
 func (s *source) Apply(c *client.Change) error {
+	i, found := metadata.Types.Lookup(c.TypeURL)
+	if !found {
+		return fmt.Errorf("invalid type: %v", c.TypeURL)
+	}
+	scope.Debugf("received object %+v and length %d", c, len(c.Objects))
 	var errs error
+	ea := make([]resource.Event, 0)
 	for _, o := range c.Objects {
+		if o.TypeURL != c.TypeURL {
+			errs = multierror.Append(errs,
+				fmt.Errorf("type %v mismatch in received object: %v", c.TypeURL, o.TypeURL))
+			continue
+		}
 		e, err := toEntry(o)
 		if err != nil {
 			errs = multierror.Append(errs, err)
+			continue
 		}
+		// If there is an error, ignore the entire batch.
+		// but keep appending errs[] for the batch
+		if errs == nil {
+			ea = append(ea, resource.Event{
+				Kind:  resource.Added,
+				Entry: e})
+		}
+	}
 
+	// At this point we have all resources for that TypeURL
+	if errs == nil {
```

Comment:
Rather than this delete->add->sync flow, can we instead add a map from `typeURL -> name -> seen` (i.e. `map[string]map[string]bool`) and use that to compute the correct events? Where:
```
for each type URL
   if an entry for this name exists, emit an `Update`
   if there is no entry, create one, emit `Add`
   if entry appears but is not in change, delete entry, emit `Delete`
```

In other words, keep enough info in hand to do the diff without holding on to the state of the objects themselves.